pkg/flow/internal/controller/loader.go

```diff
@@ -464,17 +512,57 @@ func (l *Loader) populateComponentNodes(g *dag.Graph, componentBlocks []*ast.Blo
 				})
 				continue
 			}
-
-			// Create a new component
-			c = NewComponentNode(l.globals, registration, block)
+			firstPart := strings.Split(componentName, ".")[0]
+			if l.shouldAddDeclareComponentNode(firstPart, componentName) {
+				g.Add(NewDeclareComponentNode(l.globals, block, l.getModuleInfo))
+			} else {
+				registration, exists := l.componentReg.Get(componentName)
+				if !exists {
+					diags.Add(diag.Diagnostic{
+						Severity: diag.SeverityLevelError,
+						Message:  fmt.Sprintf("Unrecognized component name %q", componentName),
+						StartPos: block.NamePos.Position(),
+						EndPos:   block.NamePos.Add(len(componentName) - 1).Position(),
+					})
+					continue
+				}
+				g.Add(NewNativeComponentNode(l.globals, registration, block))
+			}
 		}
-
-		g.Add(c)
 	}
 
 	return diags
 }
 
+func (l *Loader) shouldAddDeclareComponentNode(firstPart, componentName string) bool {
+	_, declareExists := l.declareNodes[firstPart]
+	_, importExists := l.importNodes[firstPart]
+	_, moduleDepExists := l.parentModuleDefinitions[componentName]
```

Comment:
Do we need this? I thought that we don't need to support instantiating components that were declared in parent modules?

Reply:
We do need it. One example (there are other scenarios):

```
import A {}

declare b { A.a {} }

b {}
```

In this case you need to pass the additional declare "A.a" to b.

Reply:
Right. I guess you wouldn't really call `A.a` a parent module here though? 
Let's make sure that the naming convention we adopt works here without introducing more confusion.

Reply:
In the example, the root controller imports the declare and passes it down to the child controller spawn by the custom component b. I think parent is ok but not so good because it can be confusing if you don't have the full context. That's why I remaned it to additionalDeclareContents which is provided via LoaderConfigOptions.